cohttp-eio/src/body.ml

```diff
@@ -43,56 +43,78 @@ let pp_chunk fmt = function
   | Last_chunk extensions -> pp_chunk_extension fmt extensions
 
 open Reader
+open Eio.Buf_read
 
 let read_fixed t headers =
   match Http.Header.get headers "Content-length" with
   | Some v ->
       let content_length = int_of_string v in
-      let content = take_bytes content_length t in
+      let content = take content_length t in
       content
   | None -> raise @@ Invalid_argument "Request is not a fixed content body"
 
 (* Chunked encoding parser *)
 
+open Eio.Buf_read.Syntax
+
 let hex_digit = function
   | '0' .. '9' -> true
   | 'a' .. 'f' -> true
   | 'A' .. 'F' -> true
   | _ -> false
 
-let quoted_pair =
-  char '\\'
-  *> satisfy (function ' ' | '\t' | '\x21' .. '\x7E' -> true | _ -> false)
+let quoted_char =
+  let+ c = any_char in
+  match c with
+  | ' ' | '\t' | '\x21' .. '\x7E' -> c
+  | c -> failwith (Printf.sprintf "Invalid escape \\%C" c)
 
 (*-- qdtext = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text -- *)
 let qdtext =
-  satisfy (function
-    | '\t' | ' ' | '\x21' | '\x23' .. '\x5B' -> true
-    | '\x5D' .. '\x7E' -> true
-    | _ -> false)
+  let+ c = any_char in
+  match c with
+  | '\t' | ' ' | '\x21' | '\x23' .. '\x5B'
+  | '\x5D' .. '\x7E' -> c
+  | c -> failwith (Printf.sprintf "Invalid quoted character %C" c)
 
 (*-- quoted-string = DQUOTE *( qdtext / quoted-pair ) DQUOTE --*)
-let quoted_string =
-  let dquote = char '"' in
-  let+ chars = dquote *> many_till (qdtext <|> quoted_pair) dquote <* dquote in
-  String.of_seq @@ List.to_seq chars
+let quoted_string r =
+  char '"' r;
+  let buf = Buffer.create 100 in
```

Comment:
It would be lovely if this buffer could be reused somehow by the parsing fibre...

Reply:
It's not very difficult, but we only call this once per chunk so it has no noticeable performance impact and it makes the code a little harder to read.